File: python/code/12-furrier-transform.py
import numpy as np
import matplotlib.pyplot as plt

# Sampling frequency
fs = 2000

# Time
t = np.arange(0, 1, 1 / fs)

# Given signal
x = (
    np.cos(2 * np.pi * 100 * t)
    + np.cos(2 * np.pi * 500 * t)
    + np.cos(2 * np.pi * 700 * t)
)

# Number of samples
N = len(x)

# FFT
X = np.fft.fft(x)

# Frequency
freq = np.fft.fftfreq(N, 1 / fs)

# Positive frequency
positive = (freq >= 0) & (freq <= 900)

# Amplitude spectrum
amplitude = 2 * np.abs(X) / N

# -------------------------
# Approximate Fourier integral
# -------------------------
f_values = np.linspace(0, 900, 901)

fourier_transform = np.zeros(len(f_values), dtype=complex)

dt = 1 / fs

for i, f in enumerate(f_values):
    fourier_transform[i] = np.sum(
        x * np.exp(-2j * np.pi * f * t)
    ) * dt


# -------------------------
# Plot
# -------------------------
plt.figure(figsize=(12, 8))

plt.subplot(2, 1, 1)

plt.plot(
    freq[positive],
    amplitude[positive]
)

plt.title("Amplitude Spectrum")
plt.xlabel("Frequency (Hz)")
plt.ylabel("Amplitude")
plt.grid(True)


plt.subplot(2, 1, 2)

plt.plot(
    f_values,
    np.abs(fourier_transform)
)

plt.title("Approximate Fourier Transform")
plt.xlabel("Frequency (Hz)")
plt.ylabel("|X(f)|")
plt.grid(True)

plt.tight_layout()
plt.show()

File: matlab/code/12-furrier-transform.m
clc;
clear;
close all;

% Sampling frequency
Fs = 5000;

% Duration
T = 1;

% Time vector
t = 0:1/Fs:T-1/Fs;

% Given multi-frequency signal
x = cos(2*pi*100*t) ...
    + cos(2*pi*500*t) ...
    + cos(2*pi*700*t);

% Number of samples
N = length(x);

% FFT
X = fft(x);

% Frequency vector
f_fft = (0:N-1) * Fs/N;

% Amplitude spectrum
X_mag = abs(X)/N;

% Plot time-domain signal
figure;
plot(t, x);
grid on;

xlabel('Time (s)');
ylabel('Amplitude');
title('Multi-Frequency Signal');

xlim([0 0.05]);

% Plot amplitude spectrum
figure;
plot(f_fft(1:N/2), 2*X_mag(1:N/2));
grid on;

xlabel('Frequency (Hz)');
ylabel('Amplitude');
title('Amplitude Spectrum');

xlim([0 900]);

% ------------------------------------------------
% Approximate Fourier Transform Integral
% ------------------------------------------------

% Frequency range
f = 0:1:900;

% Initialize Fourier Transform
X_integral = zeros(size(f));

% Numerical integration
for i = 1:length(f)

    X_integral(i) = trapz( ...
        t, ...
        x .* exp(-1i*2*pi*f(i)*t) ...
        );

end

% Magnitude of Fourier Transform
X_integral_mag = abs(X_integral);

% Plot approximate Fourier transform
figure;
plot(f, X_integral_mag);
grid on;

xlabel('Frequency (Hz)');
ylabel('|X(f)|');

title('Approximate Fourier Transform using Numerical Integration');

xlim([0 900]);